### ISSUE TITLE
Bug: "[ws] [Errno 32] Broken pipe" makes webmapper crash

### ISSUE THREAD

**Issue by sofian** [radarsat1/webmapper#1]
Hi,

I'm experiencing a problem which I didn't have before with webmapper. At startup it is fine. However as soon as I launch some libmapper programs (eg. influence and qualiaAgent) it crashes with the following errors:

http://pastebin.com/v6tqpEmS

**Comment by sofian** [comment]
I'm running python 2.7 on mac os x 10.6.

**Comment by radarsat1** [comment]
I've been getting this on a windows machine too. Annoying!  On this machine anyways it seems to be related to a failure in the websocket protocol.  (Not surprising.. it is still evolving I guess.)

I got it working by disabling websockets, added a "return" to the beginning of "open_websocket" in "command.js".  Can you tell me if that fixes it for you?  I'll see if I can update the websocket stuff, or maybe switch to using a proper library for it.  At the least I can make it more robust to failure, and have it properly fall back to regular connections.

**Comment by sofian** [comment]
I updated but I still got the same problem.

On 12-07-05 11:32 PM, Stephen Sinclair wrote:

> I've been getting this on a windows machine too. Annoying!  On this machine anyways it seems to be related to a failure in the websocket protocol.  (Not surprising.. it is still evolving I guess.)
> 
> I got it working by disabling websockets, added a "return" to the beginning of "open_websocket" in "command.js".  Can you tell me if that fixes it for you?  I'll see if I can update the websocket stuff, or maybe switch to using a proper library for it.  At the least I can make it more robust to failure, and have it properly fall back to regular connections.
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/radarsat1/webmapper/issues/1#issuecomment-6791225

## 

Sofian Audry
http://sofianaudry.com

**Comment by radarsat1** [comment]
Oh I just meant to see if you could change the line, I didn't actually
commit a fix yet, because I don't know if it will fix it!  Can you try
disabling websockets in webmapper_http_server.py?

On Fri, Jul 6, 2012 at 11:37 AM, sofian
reply@reply.github.com
wrote:

**Comment by sofian** [comment]
I don't know exactly what you meant by "disabling websockets". I tried
to comment out the place where do_websocket_x() was called:

Line 110-120
    def do_websocket(self):
        ref.inc()
        ws_version = self.websocket_handshake()
        try:
          print "Websockets disabled for now"

# if ws_version < 8:

# self.do_websocket_0()

# else:

# self.do_websocket_8()

```
    except socket.error, e:
        if e.errno == errno.EPIPE:
```

It still doesn't work. Output is:
[] # webmapper.py --port 8000 --no-browser
serving at port 8000
Websockets disabled for now
Websockets disabled for now
Websockets disabled for now
Websockets disabled for now
Lost connection.
shutting down...
Websockets disabled for now
bye.
[] #

On 12-07-07 7:07 AM, Stephen Sinclair wrote:

> Oh I just meant to see if you could change the line, I didn't actually
> commit a fix yet, because I don't know if it will fix it!  Can you try
> disabling websockets in webmapper_http_server.py?
> 
> On Fri, Jul 6, 2012 at 11:37 AM, sofian
> reply@reply.github.com
> wrote:
> 
> > I updated but I still got the same problem.
> > 
> > On 12-07-05 11:32 PM, Stephen Sinclair wrote:
> > 
> > > I've been getting this on a windows machine too. Annoying!  On this machine anyways it seems to be related to a failure in the websocket protocol.  (Not surprising.. it is still evolving I guess.)
> > > 
> > > I got it working by disabling websockets, added a "return" to the beginning of "open_websocket" in "command.js".  Can you tell me if that fixes it for you?  I'll see if I can update the websocket stuff, or maybe switch to using a proper library for it.  At the least I can make it more robust to failure, and have it properly fall back to regular connections.
> > > 
> > > ---
> > > 
> > > Reply to this email directly or view it on GitHub:
> > > https://github.com/radarsat1/webmapper/issues/1#issuecomment-6791225
> > 
> > ## 
> > 
> > Sofian Audry
> > http://sofianaudry.com
> > 
> > ---
> > 
> > Reply to this email directly or view it on GitHub:
> > 
> > ## https://github.com/radarsat1/webmapper/issues/1#issuecomment-6808325
> > 
> > Reply to this email directly or view it on GitHub:
> > https://github.com/radarsat1/webmapper/issues/1#issuecomment-6820703

## 

Sofian Audry
http://sofianaudry.com

**Comment by radarsat1** [comment]
Okay I think I have fixed the issue.  There was a small problem in the header response (Sec-WebSocket-Protocol) which the new version of chrome is more strict about.  Also, I made it fail a little more reasonably if websockets cause problems.  It tries only 10 times to connect to the websocket, then gives up and goes back to long polling.  It seems there's still a problem in Safari, but this still uses an old version of websockets so I think this will problem go away after some future upgrade.  In the meantime the fallback to long polling should do the trick.
